packages/dds/merge-tree/src/test/client.applyStashedOpFarm.spec.ts

```diff
@@ -0,0 +1,192 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import random from "random-js";
+import { ISequencedDocumentMessage } from "@fluidframework/protocol-definitions";
+import { IMergeTreeOp } from "../ops";
+import { SegmentGroup } from "../mergeTree";
+import {
+    generateClientNames,
+    doOverRange,
+    runMergeTreeOperationRunner,
+    annotateRange,
+    removeRange,
+    IMergeTreeOperationRunnerConfig,
+    IConfigRange,
+} from "./mergeTreeOperationRunner";
+import { TestClient } from "./testClient";
+import { TestClientLogger } from "./testClientLogger";
+
+// This test is based on reconnectFarm, but we keep a second set of clients. For
+// these clients, we apply the generated ops as stashed ops, then regenerate
+// them to simulate resubmit(), then apply them. In the end, they should arrive
+// at the same state as the "normal" set of clients
+let stashClients: TestClient[] = [];
+
+function applyMessagesWithReconnect(
+    startingSeq: number,
+    messageDatas: [ISequencedDocumentMessage, SegmentGroup | SegmentGroup[], number][],
+    clients: readonly TestClient[],
+    logger: TestClientLogger,
+) {
+    let seq = startingSeq;
+    const reconnectClientMsgs: [IMergeTreeOp, SegmentGroup | SegmentGroup[]][] = [];
+    let minSeq = 0;
+
+    // apply ops as stashed ops except for client #1
+    const stashedOps = [];
+    for (const messageData of messageDatas) {
+        if (messageData[2] !== 1) {
+            const localMetadata = stashClients[messageData[2]].applyStashedOp(messageData[0].contents);
+            stashedOps.push([messageData[0].contents, localMetadata, messageData[2]]);
+        }
+    }
+    // this should put all stash clients (except #1) in the same state as the
+    // respective normal clients, having local changes only.
+    for (let i = 0; i < clients.length; ++i) {
+        if (i !== 1) {
+            new TestClientLogger([clients[i], stashClients[i]]).validate();
```

Comment:
Can we just factor validate out of TestClientLogger into either a static or export function,  so the core logic is reusable without needing to create an instance